File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/runtime/IntervalHelperTest.java
package org.opencds.cqf.cql.engine.runtime;

import static org.junit.jupiter.api.Assertions.*;

import java.math.BigDecimal;
import java.util.Arrays;
import org.junit.jupiter.api.Test;

class IntervalHelperTest {

    @Test
    void findNonNullBoundary() {
        var intervals = Arrays.asList(null, new Interval(5, true, null, true), new Interval(null, false, 6, true));
        assertEquals(5, IntervalHelper.findNonNullBoundary(intervals));
    }

    @Test
    void quantityFromCoarsestPrecisionOfBoundaries() {
        var intervals = Arrays.asList(null, new Interval(new BigDecimal("1.12"), true, new BigDecimal("1.1234"), true));
        var quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals);
        assertEquals(1, quantity.getValue().intValue());
        assertEquals(2, quantity.getValue().scale());
        assertTrue(Quantity.isDefaultUnit(quantity.getUnit()));

        intervals = Arrays.asList(new Interval(
                null, false, new Quantity().withValue(new BigDecimal("1.123")).withUnit("g"), true));
        quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals);
        assertEquals(1, quantity.getValue().intValue());
        assertEquals(3, quantity.getValue().scale());
        assertEquals("g", quantity.getUnit());

        intervals = Arrays.asList(new Interval(new Date("2025-07-15"), true, new Date("2025-12"), true));
        quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals);
        assertEquals(1, quantity.getValue().intValue());
        assertEquals(0, quantity.getValue().scale());
        assertEquals("month", quantity.getUnit());

        intervals = Arrays.asList(new Interval(10, true, null, true));
        quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals);
        assertEquals(1, quantity.getValue().intValue());
        assertEquals(0, quantity.getValue().scale());
        assertTrue(Quantity.isDefaultUnit(quantity.getUnit()));
    }

    @Test
    void isQuantityCompatibleWithBoundaries() {
        var quantityWithDefaultUnit =
                new Quantity().withValue(new BigDecimal("1.123")).withDefaultUnit();
        var gramsQuantity = new Quantity().withValue(new BigDecimal("1.123")).withUnit("g");
        var monthsQuantity = new Quantity().withValue(new BigDecimal("10")).withUnit("month");

        var intervals = Arrays.asList(new Interval(10, true, null, false));
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals));
        assertFalse(IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals));

        intervals = Arrays.asList(new Interval(1L, true, null, true));
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals));
        assertFalse(IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals));

        intervals = Arrays.asList(null, new Interval(new BigDecimal("1.1"), true, null, true));
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals));
        assertFalse(IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals));

        intervals = Arrays.asList(
                new Interval(new Quantity().withValue(new BigDecimal("1.1")).withUnit("g"), true, null, true));
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals));
        assertFalse(IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals));

        intervals = Arrays.asList(new Interval(null, false, new Date("2025-12"), true));
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(monthsQuantity, intervals));
        assertFalse(IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals));
        assertFalse(IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals));

        intervals = Arrays.asList((Interval) null);
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals));
        assertTrue(IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals));
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/runtime/IntervalHelperTest.kt
package org.opencds.cqf.cql.engine.runtime

import java.math.BigDecimal
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class IntervalHelperTest {
    @Test
    fun findNonNullBoundary() {
        val intervals = listOf(null, Interval(5, true, null, true), Interval(null, false, 6, true))
        Assertions.assertEquals(5, IntervalHelper.findNonNullBoundary(intervals))
    }

    @Test
    fun quantityFromCoarsestPrecisionOfBoundaries() {
        var intervals = listOf(null, Interval(BigDecimal("1.12"), true, BigDecimal("1.1234"), true))
        var quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals)
        Assertions.assertEquals(1, quantity.value!!.toInt())
        Assertions.assertEquals(2, quantity.value!!.scale())
        Assertions.assertTrue(Quantity.isDefaultUnit(quantity.unit))

        intervals =
            listOf<Interval?>(
                Interval(null, false, Quantity().withValue(BigDecimal("1.123")).withUnit("g"), true)
            )
        quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals)
        Assertions.assertEquals(1, quantity.value!!.toInt())
        Assertions.assertEquals(3, quantity.value!!.scale())
        Assertions.assertEquals("g", quantity.unit)

        intervals = listOf<Interval?>(Interval(Date("2025-07-15"), true, Date("2025-12"), true))
        quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals)
        Assertions.assertEquals(1, quantity.value!!.toInt())
        Assertions.assertEquals(0, quantity.value!!.scale())
        Assertions.assertEquals("month", quantity.unit)

        intervals = listOf<Interval?>(Interval(10, true, null, true))
        quantity = IntervalHelper.quantityFromCoarsestPrecisionOfBoundaries(intervals)
        Assertions.assertEquals(1, quantity.value!!.toInt())
        Assertions.assertEquals(0, quantity.value!!.scale())
        Assertions.assertTrue(Quantity.isDefaultUnit(quantity.unit))
    }

    @Test
    fun isQuantityCompatibleWithBoundaries() {
        val quantityWithDefaultUnit = Quantity().withValue(BigDecimal("1.123")).withDefaultUnit()
        val gramsQuantity = Quantity().withValue(BigDecimal("1.123")).withUnit("g")
        val monthsQuantity = Quantity().withValue(BigDecimal("10")).withUnit("month")

        var intervals = listOf<Interval?>(Interval(10, true, null, false))
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals)
        )
        Assertions.assertFalse(
            IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals)
        )

        intervals = listOf<Interval?>(Interval(1L, true, null, true))
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals)
        )
        Assertions.assertFalse(
            IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals)
        )

        intervals = listOf(null, Interval(BigDecimal("1.1"), true, null, true))
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals)
        )
        Assertions.assertFalse(
            IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals)
        )

        intervals =
            listOf<Interval?>(
                Interval(Quantity().withValue(BigDecimal("1.1")).withUnit("g"), true, null, true)
            )
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals)
        )
        Assertions.assertFalse(
            IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals)
        )

        intervals = listOf<Interval?>(Interval(null, false, Date("2025-12"), true))
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(monthsQuantity, intervals)
        )
        Assertions.assertFalse(
            IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals)
        )
        Assertions.assertFalse(
            IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals)
        )

        intervals = listOf(null as Interval?)
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(quantityWithDefaultUnit, intervals)
        )
        Assertions.assertTrue(
            IntervalHelper.isQuantityCompatibleWithBoundaries(gramsQuantity, intervals)
        )
    }
}
